### ISSUE TITLE
Setting -DARROW_HDFS=OFF breaks arrow build when linking against boost libraries

### ISSUE THREAD

**Issue by pearu** [apache/arrow#2316]
When using `cmake -DARROW_HDFS=OFF ..` then boost libraries are not detected (ubuntu 16.04) and the arrow build fails:
```
...
Scanning dependencies of target table_builder-test
[ 72%] Building CXX object src/arrow/CMakeFiles/table_builder-test.dir/table_builder-test.cc.o
[ 72%] Linking CXX executable ../../debug/table_builder-test
../../debug/libarrow.a(decimal.cc.o): In function `bool boost::regex_match<__gnu_cxx::__normal_iterator<char const*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<boost::sub_match<__gnu_cxx::__normal_iterator<char const*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > >, char, boost::regex_traits<char, boost::cpp_regex_traits<char> > >(__gnu_cxx::__normal_iterator<char const*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, __gnu_cxx::__normal_iterator<char const*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, boost::match_results<__gnu_cxx::__normal_iterator<char const*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<boost::sub_match<__gnu_cxx::__normal_iterator<char const*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > > >&, boost::basic_regex<char, boost::regex_traits<char, boost::cpp_regex_traits<char> > > const&, boost::regex_constants::_match_flags)':
/usr/include/boost/regex/v4/regex_match.hpp:50: undefined reference to `boost::re_detail::perl_matcher<__gnu_cxx::__normal_iterator<char const*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<boost::sub_match<__gnu_cxx::__normal_iterator<char const*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > >, boost::regex_traits<char, boost::cpp_regex_traits<char> > >::match()'
...
```

With `ARROW_HDFS=ON`, the build succeeds.

The problematic code appears to be
https://github.com/apache/arrow/blob/078b8068f1c7176ff4a56aa95c9b57800aacce82/cpp/CMakeLists.txt#L307

Removing the `else` block with `ARROW_BOOST_HEADER_ONLY=1` makes `cmake -DARROW_HDFS=OFF ..` successful.

**Comment by wesm** [comment]
hi @pearu, can you open a JIRA for this? Would you be up to submit a patch? https://github.com/apache/arrow#how-to-contribute

**Comment by wesm** [comment]
Thanks! This got broken when we added decimal support (which uses boost::regex, non-header-only) a while back. Will see that it's fixed for 0.10.0
